core/runner/src/test/java/com/cognifide/aet/runner/distribution/dispatch/CollectorJobSchedulerTest.java

```diff
@@ -95,17 +95,19 @@ public void add_expectMessageSent() throws Exception {
     verify(producer, times(1)).send(Matchers.<Destination>any(), Matchers.<Message>any());
   }
 
-
-  @Ignore("Ignored for some unexpected test failure on Linux")
-  @Test(timeout = 2000l, expected = IllegalStateException.class)
-  public void add_whenAddedMessageExists_expectIllegalStateException() throws Exception {
+  @Test
+  public void add_messageQueueIsConsumedBySafeRunMethod() throws Exception {
     String correlationID = "98765432100";
+    Queue<MessageWithDestination> messagesQueue1 = mockMessagesQueue(Mockito.mock(Message.class),
+            Mockito.mock(Destination.class), correlationID);
+    Queue<MessageWithDestination> messagesQueue2 = mockMessagesQueue(Mockito.mock(Message.class),
+            Mockito.mock(Destination.class), correlationID);
 
-    Queue<MessageWithDestination> messagesQueue1 = mockMessagesQueue(Mockito.mock(Message.class), Mockito.mock
-            (Destination.class), correlationID);
-    Queue<MessageWithDestination> messagesQueue2 = mockMessagesQueue(Mockito.mock(Message.class), Mockito.mock
-            (Destination.class), correlationID);
     tested.add(messagesQueue1, correlationID);
+    // messageQueues will be processed and removed from messagesMap
+    // in safeRun method of CollectorJobScheduler
+    // so it is possible to add another with the same correlationId
+    Thread.sleep(1000);
```

Comment:
Hi @Skejven , I had a closer look into the test and the class and need to consult something:

the flow that i debugged on Windows with IntelliJ is as follows:

1. wait for `availableMessages` semaphore
2. *messege queue 1* is added to `messagesMap`
3. *message queue 1* is processed in `safeRun` method and sent
4. wait for `availableMessages` semaphore
5. `safeRun` method checks that the *message queue 1* has no more entries
6. *message queue 1* is removed from `messagesMap`

There may be a race condition. If the `safeRun` method is quick enough to make two loops, then it consumes the *message queue 1* and also removes it from `messageMap`. Otherwise (if only one loop happens) there will be exception thrown while adding second message queue with same correlation ID.

As this issue is complicated, could we have someone to review this?
I have set breakpoints at the two lines where the messege queues are added and also in `CollectorJobSchedule`: lines 255, 129.

I will add another pull request with proposed solution...